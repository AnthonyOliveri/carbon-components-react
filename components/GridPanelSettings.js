import React from 'react';
import PropTypes from 'prop-types';
import Icon from './Icon';
import OverflowMenu from './OverflowMenu';
import OverflowMenuItem from './OverflowMenuItem';

export default function GridPanelSettings(props) {
  const renderHideableColumns = (items, toggleColumnVisibility) =>
    items.map((item, key) => {
      const content = (
        <div className="bx--grid-panel-settings__item">
          <Icon
            name="checkmark"
            className={
              item.hidden ? 'bx--grid-panel-settings__icon-hidden' : ''
            }
            description="The column is visible"
          />
          {item.title}
        </div>
      );

      return (
        <OverflowMenuItem
          className="bx--grid-panel-settings__toggle"
          onClick={() => toggleColumnVisibility(item)}
          itemText={content}
          key={key}
        />
      );
    });

  return (
    <OverflowMenu className="bx--grid-panel-settings">
      <div className="label">
        Columns
      </div>
      {renderHideableColumns(props.items, props.toggleColumnVisibility)}
      <hr />
      <OverflowMenuItem
        className="bx--grid-panel-settings__reset"
        onClick={props.resetColumns}
        itemText="Reset Columns to Default"
        key="reset-column-button"
      />
    </OverflowMenu>
  );
}

GridPanelSettings.propTypes = {
  items: PropTypes.array.isRequired,
  resetColumns: PropTypes.func,
  toggleColumnVisibility: PropTypes.func,
};
