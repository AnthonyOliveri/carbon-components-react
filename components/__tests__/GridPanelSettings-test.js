import GridPanelSettings from '../GridPanelSettings';
import React from 'react';
import { shallow } from 'enzyme';

jest.unmock('../GridPanelSettings');

const shallowRender = props => shallow(<GridPanelSettings {...props} />);

describe('GridPanelSettings', () => {
  describe('check if props are rendered correctly', () => {
    describe('initial render', () => {
      const wrapper = shallowRender({
        items: [
          {
            name: 'column3',
            title: 'column3',
            hidden: true,
          },
          {
            name: 'column4',
            title: 'column4',
            hidden: false,
          },
        ],
        resetColumns: jest.fn(),
        toggleColumnVisibility: jest.fn(),
      });

      it('renders correctly', () => {
        expect(wrapper).toMatchSnapshot();
      });
    });

    describe('Menu open state', () => {
      const wrapper = shallowRender({
        items: [
          {
            name: 'column6',
            title: 'column6',
            hidden: false,
          },
          {
            name: 'column8',
            title: 'column8',
            hidden: true,
          },
        ],
        resetColumns: jest.fn(),
        toggleColumnVisibility: jest.fn(),
      });

      wrapper.find('.bx--grid-panel-settings').simulate('click', {
        stopPropagation: jest.fn(),
      });
      wrapper.update();

      it('menu open renders correctly', () => {
        expect(wrapper).toMatchSnapshot();
      });
    });

    describe('Menu close state', () => {
      const wrapper = shallowRender({
        items: [
          {
            name: 'column3',
            title: 'column3',
            hidden: true,
          },
          {
            name: 'column4',
            title: 'column4',
            hidden: false,
          },
        ],
        resetColumns: jest.fn(),
        toggleColumnVisibility: jest.fn(),
      });

      // First click to open the menu
      wrapper.find('.bx--grid-panel-settings').simulate('click', {
        stopPropagation: jest.fn(),
      });

      // Second click to close the menu - the actual test
      wrapper.find('.bx--grid-panel-settings').simulate('click', {
        stopPropagation: jest.fn(),
      });

      it('menu close renders correctly', () => {
        expect(wrapper).toMatchSnapshot();
      });
    });
  });

  describe('check if events are triggered properly', () => {
    const resetColumnsMock = jest.fn();
    const toggleColumnVisibilityMock = jest.fn();
    const wrapper = shallowRender({
      items: [
        {
          name: 'column3',
          title: 'column3',
          hidden: true,
        },
        {
          name: 'column4',
          title: 'column4',
          hidden: false,
        },
      ],
      resetColumns: resetColumnsMock,
      toggleColumnVisibility: toggleColumnVisibilityMock,
    });

    // Open the menu before clicking the buttons
    wrapper.find('.bx--grid-panel-settings').simulate('click', {
      stopPropagation: jest.fn(),
    });

    it('resetColumns is called', () => {
      wrapper.find('.bx--grid-panel-settings__reset').simulate('click');
      expect(resetColumnsMock).toBeCalled();
    });

    it('toggleColumnVisibility is called', () => {
      wrapper.find('.bx--grid-panel-settings__toggle').at(0).simulate('click');
      expect(toggleColumnVisibilityMock).toBeCalledWith({
        name: 'column3',
        title: 'column3',
        hidden: true,
      });
    });
  });
});
