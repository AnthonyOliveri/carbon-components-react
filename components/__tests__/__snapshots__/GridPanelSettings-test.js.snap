exports[`GridPanelSettings check if props are rendered correctly Menu close state menu close renders correctly 1`] = `
<OverflowMenu
  ariaLabel="list of options"
  className="bx--grid-panel-settings"
  flipped={false}
  floatingMenu={false}
  iconDescription="open and close list of options"
  iconName="overflow-menu"
  menuOffset={
    Object {
      "left": 60.5,
      "top": 0,
    }
  }
  menuOffsetFlip={
    Object {
      "left": -60.5,
      "top": 0,
    }
  }
  onClick={[Function]}
  open={false}
  tabIndex={0}>
  <div
    className="label">
    Columns
  </div>
  <OverflowMenuItem
    className="bx--grid-panel-settings__toggle"
    hasDivider={false}
    isDelete={false}
    itemText={
      <div
        className="bx--grid-panel-settings__item">
        <Icon
          className="bx--grid-panel-settings__icon-hidden"
          description="The column is visible"
          fillRule="evenodd"
          name="checkmark"
          role="img" />
        column3
      </div>
    }
    onClick={[Function]} />
  <OverflowMenuItem
    className="bx--grid-panel-settings__toggle"
    hasDivider={false}
    isDelete={false}
    itemText={
      <div
        className="bx--grid-panel-settings__item">
        <Icon
          className=""
          description="The column is visible"
          fillRule="evenodd"
          name="checkmark"
          role="img" />
        column4
      </div>
    }
    onClick={[Function]} />
  <hr />
  <OverflowMenuItem
    className="bx--grid-panel-settings__reset"
    hasDivider={false}
    isDelete={false}
    itemText="Reset Columns to Default"
    onClick={[Function]} />
</OverflowMenu>
`;

exports[`GridPanelSettings check if props are rendered correctly Menu open state menu open renders correctly 1`] = `
<OverflowMenu
  ariaLabel="list of options"
  className="bx--grid-panel-settings"
  flipped={false}
  floatingMenu={false}
  iconDescription="open and close list of options"
  iconName="overflow-menu"
  menuOffset={
    Object {
      "left": 60.5,
      "top": 0,
    }
  }
  menuOffsetFlip={
    Object {
      "left": -60.5,
      "top": 0,
    }
  }
  onClick={[Function]}
  open={false}
  tabIndex={0}>
  <div
    className="label">
    Columns
  </div>
  <OverflowMenuItem
    className="bx--grid-panel-settings__toggle"
    hasDivider={false}
    isDelete={false}
    itemText={
      <div
        className="bx--grid-panel-settings__item">
        <Icon
          className=""
          description="The column is visible"
          fillRule="evenodd"
          name="checkmark"
          role="img" />
        column6
      </div>
    }
    onClick={[Function]} />
  <OverflowMenuItem
    className="bx--grid-panel-settings__toggle"
    hasDivider={false}
    isDelete={false}
    itemText={
      <div
        className="bx--grid-panel-settings__item">
        <Icon
          className="bx--grid-panel-settings__icon-hidden"
          description="The column is visible"
          fillRule="evenodd"
          name="checkmark"
          role="img" />
        column8
      </div>
    }
    onClick={[Function]} />
  <hr />
  <OverflowMenuItem
    className="bx--grid-panel-settings__reset"
    hasDivider={false}
    isDelete={false}
    itemText="Reset Columns to Default"
    onClick={[Function]} />
</OverflowMenu>
`;

exports[`GridPanelSettings check if props are rendered correctly initial render renders correctly 1`] = `
<OverflowMenu
  ariaLabel="list of options"
  className="bx--grid-panel-settings"
  flipped={false}
  floatingMenu={false}
  iconDescription="open and close list of options"
  iconName="overflow-menu"
  menuOffset={
    Object {
      "left": 60.5,
      "top": 0,
    }
  }
  menuOffsetFlip={
    Object {
      "left": -60.5,
      "top": 0,
    }
  }
  onClick={[Function]}
  open={false}
  tabIndex={0}>
  <div
    className="label">
    Columns
  </div>
  <OverflowMenuItem
    className="bx--grid-panel-settings__toggle"
    hasDivider={false}
    isDelete={false}
    itemText={
      <div
        className="bx--grid-panel-settings__item">
        <Icon
          className="bx--grid-panel-settings__icon-hidden"
          description="The column is visible"
          fillRule="evenodd"
          name="checkmark"
          role="img" />
        column3
      </div>
    }
    onClick={[Function]} />
  <OverflowMenuItem
    className="bx--grid-panel-settings__toggle"
    hasDivider={false}
    isDelete={false}
    itemText={
      <div
        className="bx--grid-panel-settings__item">
        <Icon
          className=""
          description="The column is visible"
          fillRule="evenodd"
          name="checkmark"
          role="img" />
        column4
      </div>
    }
    onClick={[Function]} />
  <hr />
  <OverflowMenuItem
    className="bx--grid-panel-settings__reset"
    hasDivider={false}
    isDelete={false}
    itemText="Reset Columns to Default"
    onClick={[Function]} />
</OverflowMenu>
`;
